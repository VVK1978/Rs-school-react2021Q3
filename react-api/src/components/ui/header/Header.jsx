import React from 'react';
import Search from '../search/Search.jsx';

export default function Header(props) {
  const { data } = props;
  const { updateState } = props;
  return (
    <header className="head">
      <Search data={data} updateState={updateState} />
    </header>
  );
}
