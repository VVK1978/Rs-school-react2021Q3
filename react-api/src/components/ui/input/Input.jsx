import React from 'react';

export default function Input() {
  const handlerChange = () => {};
  return (
    <input
      type="text"
      className="search-input"
      onChange={handlerChange}
      name="search"
      placeholder="Search by name..."
    />
  );
}
