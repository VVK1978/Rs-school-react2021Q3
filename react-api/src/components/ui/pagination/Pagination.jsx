import React from 'react';
import getSearchCharacters from '../../../services/api.jsx';
import pageCreator from '../../utils/pagesCreator';

export default function Pagination(props) {
  const { data } = props;
  const { updateState } = props;
  const pages = pageCreator(data.totalPages, data.perPage, data.currentPage);
  const handleClick = async (page) => {
    const currentPage = page;
    data.currentPage = page;
    const response = await getSearchCharacters(data);
    const characters = response.docs;
    const updatedData = await { ...data, characters, currentPage };
    await updateState(updatedData);
  };
  return (
    <>
      {data.characters.length !== 0 ? (
        <>
          <div className="pagination-container">
            {pages.map((page, index) => (
              <div key={index}>
                <span
                  className={`pagination-page ${
                    data.currentPage === page ? 'active' : ''
                  }`}
                  onClick={() => handleClick(page)}
                >
                  {page}
                </span>
              </div>
            ))}
          </div>
        </>
      ) : (
        ''
      )}
    </>
  );
}
