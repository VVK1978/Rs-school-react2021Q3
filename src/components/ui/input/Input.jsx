import React from 'react';
import { useDispatch } from 'react-redux';
import { setSearchValue } from '../../reducers/characterReducer';

export default function Input() {
  const dispatch = useDispatch();

  const handlerChange = (event) => {
    dispatch(setSearchValue(event.target.value));
  };
  return (
    <>
      <input
        type="text"
        className="search-input"
        onChange={handlerChange}
				name="search"
				placeholder='Search by name...'
      />
    </>
  );
}
