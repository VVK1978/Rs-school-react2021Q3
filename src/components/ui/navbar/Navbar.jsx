import React from 'react';
import { NavLink } from 'react-router-dom';

export default function Navbar() {
  return (
    <>
      <nav className="nav-content">
        <ul className="nav-menu">
          <li className="menu-item">
            <NavLink exact to="/" className="menu-item">
              Home
            </NavLink>
          </li>
          <li className="menu-item">
            <NavLink to="/about" className="menu-item">
              About
            </NavLink>
          </li>
        </ul>
      </nav>
    </>
  );
}
