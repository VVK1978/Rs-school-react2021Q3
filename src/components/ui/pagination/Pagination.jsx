import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { setCurrentPage } from '../../reducers/characterReducer';
import { getSearchCharacters } from '../../../actions/characters.jsx';
import pageCreator from '../../utils/pagesCreator';

export default function Pagination() {
  const totalPages = useSelector((state) => state.characters.totalPages);
  const perPage = useSelector((state) => state.characters.perPage);
  const currentPage = useSelector((state) => state.characters.currentPage);
  const searchValue = useSelector((state) => state.characters.searchValue);
  const sortBy = useSelector((state) => state.characters.sortBy);
  const dispatch = useDispatch();
  const pages = pageCreator(totalPages, perPage, currentPage);
  const handleClick = (page) => {
    dispatch(setCurrentPage(page));
    dispatch(getSearchCharacters(searchValue, perPage, page, sortBy));
  };
  return (
    <div className="pagination-container">
      {pages.map((page, index) => (
        <div key={index}>
          <span
            className={`pagination-page ${
              currentPage === page ? 'active' : ''
            }`}
            onClick={() => handleClick(page)}
          >
            {page}
          </span>
        </div>
      ))}
    </div>
  );
}
