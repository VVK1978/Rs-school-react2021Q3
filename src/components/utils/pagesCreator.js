export default function pagesCreator(totalPages, currentPage) {
  let pages = [];
  if (totalPages < 11) {
    pages = Array(totalPages)
      .fill()
      .map((item, ind) => ind + 1);
  } else if (currentPage > 5) {
    for (let i = currentPage - 4; i <= currentPage + 5; i += 1) {
      pages.push(i);
      if (totalPages === i) break;
    }
  } else {
    for (let i = 1; i <= 10; i += 1) {
      pages.push(i);
      if (totalPages === i) break;
    }
  }
  return pages;
}
