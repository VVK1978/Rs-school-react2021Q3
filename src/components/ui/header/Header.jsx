import React from 'react';
import Search from '../search/Search';
import Navbar from '../navbar/Navbar';

export default function Header() {
  const h1 = <h1 className="main-title">React.SSR</h1>;

  return (
    <>
      {h1}
      <header className="head">
        <Navbar />
        <Search />
      </header>
    </>
  );
}
