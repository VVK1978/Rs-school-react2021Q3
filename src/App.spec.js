/* eslint-disable import/no-duplicates */
/* eslint-disable import/no-named-as-default */
/* eslint-disable no-undef */
import shallow from 'enzyme/build/shallow';
import React from 'react';
import * as redux from 'react-redux';
import Button from './components/ui/button/button.jsx';
import Header from './components/ui/header/Header.jsx';
import Input from './components/ui/input/Input.jsx';
import Navbar from './components/ui/navbar/Navbar.jsx';
import Pagination from './components/ui/pagination/Pagination.jsx';
import pagesCreator from './components/utils/pagesCreator';
import setPerPage from './components/reducers/characterReducer';
import SET_PER_PAGE from './components/reducers/characterReducer';

const pages = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

describe('pages creator function', () => {
  it('should if pages=10', () => {
    expect(pagesCreator(10)).toEqual(pages);
  });
});

describe('pages creator function', () => {
  it('should if pages=20,current page=10', () => {
    expect(pagesCreator(20, 10)).toEqual([6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
  });
});

describe('pages creator function', () => {
  it('should if pages=20,current page=4', () => {
    expect(pagesCreator(20, 4)).toEqual(pages);
  });
});

describe('Button component', () => {
  it('should render Button component', () => {
    const component = shallow(<Button />);
    expect(component).toMatchSnapshot();
  });
});

describe('Header component', () => {
  it('should render Header component', () => {
    const component = shallow(<Header />);
    expect(component).toMatchSnapshot();
  });
});

describe('Navbar component', () => {
  it('should render Navbar component', () => {
    const component = shallow(<Navbar />);
    expect(component).toMatchSnapshot();
  });
});

describe('Navbar component', () => {
  it('should render Navbar component', () => {
    const component = shallow(<Navbar />);
    expect(component).toMatchSnapshot();
  });
});

describe('Component with redux hooks', () => {
  let spyOnUseSelector;
  let spyOnUseDispatch;
  let mockDispatch;

  beforeEach(() => {
    spyOnUseSelector = jest.spyOn(redux, 'useSelector');
    spyOnUseSelector.mockReturnValue([{ id: 1, text: 'Old Item' }]);
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
    mockDispatch = jest.fn();
    spyOnUseDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should Limit per page', () => {
    const result = setPerPage(20, SET_PER_PAGE);
    expect(result).toBe(setPerPage(20, SET_PER_PAGE));
  });

  it('should render Pagination component', () => {
    const component = shallow(<Pagination />);
    expect(component).toMatchSnapshot();
  });
});
