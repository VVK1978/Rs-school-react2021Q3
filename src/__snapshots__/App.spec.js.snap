// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render Button component 1`] = `ShallowWrapper {}`;

exports[`Component with redux hooks should render Input component 1`] = `ShallowWrapper {}`;

exports[`Component with redux hooks should render Input component 2`] = `ShallowWrapper {}`;

exports[`Component with redux hooks should render Input component 3`] = `ShallowWrapper {}`;

exports[`Component with redux hooks should render Limit component 1`] = `ShallowWrapper {}`;

exports[`Component with redux hooks should render Pagination component 1`] = `ShallowWrapper {}`;

exports[`Component with redux hooks should render Search component 1`] = `ShallowWrapper {}`;

exports[`Header component should render Header component 1`] = `ShallowWrapper {}`;

exports[`Navbar component should render Navbar component 1`] = `ShallowWrapper {}`;

exports[`Navbar component should render Navbar component 2`] = `ShallowWrapper {}`;

exports[`Pagination component should render Pagination component 1`] = `ShallowWrapper {}`;

exports[`should render characters component 1`] = `ShallowWrapper {}`;
